File: pulse-go/internal/pulse/engine/piece_test.go
/*
 * Copyright 2013-2024 Phokham Nonava
 *
 * Use of this source code is governed by the MIT license that can be
 * found in the LICENSE file.
 */

package engine

import (
	"testing"
)

func TestPiece(t *testing.T) {
	t.Run("Pieces array should be valid", func(t *testing.T) {
		for pc := range Pieces {
			if Pieces[pc] != pc {
				t.Errorf("wanted Pieces at index %d to be %v, but got %v", pc, pc, Pieces[pc])
			}
		}
	})
}

func TestPiece_IsValidPiece(t *testing.T) {
	tests := []struct {
		piece Piece
		want  bool
	}{
		{piece: WhitePawn, want: true},
		{piece: WhiteKnight, want: true},
		{piece: WhiteBishop, want: true},
		{piece: WhiteRook, want: true},
		{piece: WhiteQueen, want: true},
		{piece: WhiteKing, want: true},
		{piece: BlackPawn, want: true},
		{piece: BlackKnight, want: true},
		{piece: BlackBishop, want: true},
		{piece: BlackRook, want: true},
		{piece: BlackQueen, want: true},
		{piece: BlackKing, want: true},
		{piece: NoPiece, want: false},
	}
	for _, tt := range tests {
		t.Run("IsValidPiece should return true if piece is valid, false otherwise", func(t *testing.T) {
			valid := IsValidPiece(tt.piece)
			if valid != tt.want {
				t.Errorf("IsValidPiece(%d) = %v, want %v", tt.piece, valid, tt.want)
			}
		})
	}
}

func TestPiece_PieceOf(t *testing.T) {
	tests := []struct {
		color     Color
		pieceType PieceType
		want      Piece
	}{
		{color: White, pieceType: Pawn, want: WhitePawn},
		{color: White, pieceType: Knight, want: WhiteKnight},
		{color: White, pieceType: Bishop, want: WhiteBishop},
		{color: White, pieceType: Rook, want: WhiteRook},
		{color: White, pieceType: Queen, want: WhiteQueen},
		{color: White, pieceType: King, want: WhiteKing},
		{color: Black, pieceType: Pawn, want: BlackPawn},
		{color: Black, pieceType: Knight, want: BlackKnight},
		{color: Black, pieceType: Bishop, want: BlackBishop},
		{color: Black, pieceType: Rook, want: BlackRook},
		{color: Black, pieceType: Queen, want: BlackQueen},
		{color: Black, pieceType: King, want: BlackKing},
	}
	for _, tt := range tests {
		t.Run("PieceOf should return the piece", func(t *testing.T) {
			pc := PieceOf(tt.color, tt.pieceType)
			if pc != tt.want {
				t.Errorf("wanted piece of color %v and piece type %v to be %v, but got %v", tt.color, tt.pieceType, tt.want, pc)
			}
		})
	}
}

func TestPiece_PieceColorOf(t *testing.T) {
	tests := []struct {
		piece Piece
		want  Color
	}{
		{piece: WhitePawn, want: White},
		{piece: WhiteKnight, want: White},
		{piece: WhiteBishop, want: White},
		{piece: WhiteRook, want: White},
		{piece: WhiteQueen, want: White},
		{piece: WhiteKing, want: White},
		{piece: BlackPawn, want: Black},
		{piece: BlackKnight, want: Black},
		{piece: BlackBishop, want: Black},
		{piece: BlackRook, want: Black},
		{piece: BlackQueen, want: Black},
		{piece: BlackKing, want: Black},
	}
	for _, tt := range tests {
		t.Run("PieceColorOf should return the piece color", func(t *testing.T) {
			col := PieceColorOf(tt.piece)
			if col != tt.want {
				t.Errorf("wanted color of %v to be %v, but got %v", tt.piece, tt.want, col)
			}
		})
	}
}

func TestPiece_PieceTypeOf(t *testing.T) {
	tests := []struct {
		piece Piece
		want  PieceType
	}{
		{piece: WhitePawn, want: Pawn},
		{piece: WhiteKnight, want: Knight},
		{piece: WhiteBishop, want: Bishop},
		{piece: WhiteRook, want: Rook},
		{piece: WhiteQueen, want: Queen},
		{piece: WhiteKing, want: King},
		{piece: BlackPawn, want: Pawn},
		{piece: BlackKnight, want: Knight},
		{piece: BlackBishop, want: Bishop},
		{piece: BlackRook, want: Rook},
		{piece: BlackQueen, want: Queen},
		{piece: BlackKing, want: King},
	}
	for _, tt := range tests {
		t.Run("PieceTypeOf should return the piece type", func(t *testing.T) {
			pt := PieceTypeOf(tt.piece)
			if pt != tt.want {
				t.Errorf("wanted piece type of %v to be %v, but got %v", tt.piece, tt.want, pt)
			}
		})
	}
}

File: pulse-kotlin/src/commonTest/kotlin/com/fluxchess/pulse/kotlin/engine/PieceTest.kt
/*
 * Copyright 2013-2024 Phokham Nonava
 *
 * Use of this source code is governed by the MIT license that can be
 * found in the LICENSE file.
 */

package com.fluxchess.pulse.kotlin.engine

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertTrue

class PieceTest {
	@Test
	fun `Pieces array should be valid`() {
		for (piece in pieces.indices) {
			assertEquals(pieces[piece], piece)
		}
	}

	@Test
	fun `isValidPiece should return true if piece is valid false otherwise`() {
		assertTrue(isValidPiece(WHITE_PAWN))
		assertTrue(isValidPiece(WHITE_KNIGHT))
		assertTrue(isValidPiece(WHITE_BISHOP))
		assertTrue(isValidPiece(WHITE_ROOK))
		assertTrue(isValidPiece(WHITE_QUEEN))
		assertTrue(isValidPiece(WHITE_KING))
		assertTrue(isValidPiece(BLACK_PAWN))
		assertTrue(isValidPiece(BLACK_KNIGHT))
		assertTrue(isValidPiece(BLACK_BISHOP))
		assertTrue(isValidPiece(BLACK_ROOK))
		assertTrue(isValidPiece(BLACK_QUEEN))
		assertTrue(isValidPiece(BLACK_KING))

		assertFalse(isValidPiece(NO_PIECE))
	}

	@Test
	fun `pieceOf should return the correct piece`() {
		assertEquals(WHITE_PAWN, pieceOf(WHITE, PAWN))
		assertEquals(WHITE_KNIGHT, pieceOf(WHITE, KNIGHT))
		assertEquals(WHITE_BISHOP, pieceOf(WHITE, BISHOP))
		assertEquals(WHITE_ROOK, pieceOf(WHITE, ROOK))
		assertEquals(WHITE_QUEEN, pieceOf(WHITE, QUEEN))
		assertEquals(WHITE_KING, pieceOf(WHITE, KING))
		assertEquals(BLACK_PAWN, pieceOf(BLACK, PAWN))
		assertEquals(BLACK_KNIGHT, pieceOf(BLACK, KNIGHT))
		assertEquals(BLACK_BISHOP, pieceOf(BLACK, BISHOP))
		assertEquals(BLACK_ROOK, pieceOf(BLACK, ROOK))
		assertEquals(BLACK_QUEEN, pieceOf(BLACK, QUEEN))
		assertEquals(BLACK_KING, pieceOf(BLACK, KING))
	}

	@Test
	fun `pieceColorOf should return the piece color`() {
		assertEquals(WHITE, pieceColorOf(WHITE_PAWN))
		assertEquals(WHITE, pieceColorOf(WHITE_KNIGHT))
		assertEquals(WHITE, pieceColorOf(WHITE_BISHOP))
		assertEquals(WHITE, pieceColorOf(WHITE_ROOK))
		assertEquals(WHITE, pieceColorOf(WHITE_QUEEN))
		assertEquals(WHITE, pieceColorOf(WHITE_KING))
		assertEquals(BLACK, pieceColorOf(BLACK_PAWN))
		assertEquals(BLACK, pieceColorOf(BLACK_KNIGHT))
		assertEquals(BLACK, pieceColorOf(BLACK_BISHOP))
		assertEquals(BLACK, pieceColorOf(BLACK_ROOK))
		assertEquals(BLACK, pieceColorOf(BLACK_QUEEN))
		assertEquals(BLACK, pieceColorOf(BLACK_KING))
	}

	@Test
	fun `pieceTypeOf should return the piece type`() {
		assertEquals(PAWN, pieceTypeOf(WHITE_PAWN))
		assertEquals(KNIGHT, pieceTypeOf(WHITE_KNIGHT))
		assertEquals(BISHOP, pieceTypeOf(WHITE_BISHOP))
		assertEquals(ROOK, pieceTypeOf(WHITE_ROOK))
		assertEquals(QUEEN, pieceTypeOf(WHITE_QUEEN))
		assertEquals(KING, pieceTypeOf(WHITE_KING))
		assertEquals(PAWN, pieceTypeOf(BLACK_PAWN))
		assertEquals(KNIGHT, pieceTypeOf(BLACK_KNIGHT))
		assertEquals(BISHOP, pieceTypeOf(BLACK_BISHOP))
		assertEquals(ROOK, pieceTypeOf(BLACK_ROOK))
		assertEquals(QUEEN, pieceTypeOf(BLACK_QUEEN))
		assertEquals(KING, pieceTypeOf(BLACK_KING))
	}
}
